src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_CheckOrReachability.cs

```diff
@@ -787,6 +795,7 @@ static TypeSymbol narrowedTypeForBinary(BoundPattern resultLeft, BoundPattern re
 
             public override BoundNode? Visit(BoundNode? node)
             {
+                Debug.Assert(node is not BoundPattern { IsUnionMatching: true });
```

Comment:
Think inverting would make this clearer:
```suggestion
                Debug.Assert(node is BoundPattern { IsUnionMatching: false });
```

Reply:
Well, the very next assertion is that the visitor only handles very specific pattern nodes, so I think it would be fine to strengthen this assertion and make it more readable at the same time.